### ISSUE TITLE
Date string with fractional seconds not supported

### ISSUE THREAD

**Issue by GermaineG** [anno4j/anno4j#168]
I try to set the generation date of an annotation with the setGenerated method.
It fails with 
```
Exception in thread "main" org.openrdf.repository.object.exceptions.ObjectPersistException: Incorrect timestamp format supported. The timestamp needs to be conform to the ISO 8601 specification.
	at com.github.anno4j.model.AnnotationSupport.setGenerated(AnnotationSupport.java:275)
```

The date string is created in Java 8 with `java.time.ZonedDateTime` and `java.time.format.DateTimeFormatter`, like this:

```
ZonedDateTime now = ZonedDateTime.now();
String nowISOString = now.format(DateTimeFormatter.ISO_INSTANT);
```

A resulting date string looks like this: `2018-01-11T15:43:39.520Z`

From what I can see, in anno4j it gets tested with something basically equivalent to this:

```
org.joda.time.format.DateTimeFormatter format = org.joda.time.format.ISODateTimeFormat.dateTimeNoMillis().withZoneUTC();

try {
    format.parseDateTime(now.format(DateTimeFormatter.ISO_INSTANT));
} catch (IllegalArgumentException e) {
    ...
}
```

with fails with `java.lang.IllegalArgumentException: Invalid format: "2018-01-11T15:43:39.520Z" is malformed at ".520Z"` and leads to the Exception in anno4j above.

Is there a reason for the use of `dateTimeNoMillis()` instead of `dateTime()`?

The WADM states:

> The datetime must be a [xsd:dateTime](https://www.w3.org/TR/xmlschema-2/#dateTime) with the UTC timezone expressed as "Z".

The xsd:dateTime specification states:

> The ·lexical space· of dateTime consists of finite-length sequences of characters of the form: '-'? yyyy '-' mm '-' dd 'T' hh ':' mm ':' ss ('.' s+)? (zzzzzz)?

So imho a zulu format date string of the form `'-'? yyyy '-' mm '-' dd 'T' hh ':' mm ':' ss ('.' s+)? z` should be usable?

**Comment by berndl** [comment]
Hi @GermaineG !

Hm no, there is no real reason why we implemented it like that. I oriented myself on the examples found in the WADM, e.g. EXAMPLE 11: Lifecycle Information defines the property "created" to the value "2015-01-28T12:00:00Z", a String without milliseconds.

This String is accepted by our current implementation, while "2015-01-28T12:00:00.00Z" would not be, as you indicated.

I totally agree with you that the definition of the xsd:dateTime indicates that there should be optional milliseconds. However the implementations 

        DateTimeFormatter formatNoMillis = ISODateTimeFormat.dateTimeNoMillis().withZoneUTC();
        DateTimeFormatter formatWithMillis = ISODateTimeFormat.dateTime().withZoneUTC();

seem to produce conflicting results, although I think that one should include both possibilities of having milliseconds or not.

Long story short, I tried to combine both of them and will update our current SNAPSHOT version, so both time String representations should be accepted!

I hope this does help you.
Best regards,
Manu
